File: Java/Binary-Tree-Maximum-Path-Sum.java
/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */
public class Solution {
    /**
     * @param root: The root of binary tree.
     * @return: An integer.
     */
    private int maxv = Integer.MIN_VALUE;
    public int helper(TreeNode root) {
        if (root == null) return 0;
        int leftv = helper(root.left);
        int rightv = helper(root.right);
        int temp = Math.max(root.val, Math.max(root.val + leftv, root.val + rightv));
        maxv = Math.max(maxv, Math.max(temp, root.val + rightv + leftv));
        return temp;
    }
    public int maxPathSum(TreeNode root) {
        // write your code here
        helper(root);
        return maxv;
    }
}

File: Cpp/Binary-Tree-Maximum-Path-Sum.cpp
/**
 * Definition of TreeNode:
 */
class TreeNode {
public:
    int val;
    TreeNode *left, *right;
    TreeNode(int val) {
        this->val = val;
        this->left = this->right = NULL;
    }
};

class Solution {
public:
    /**
     * @param root: The root of binary tree.
     * @return: An integer
     */
    class ResultType {
    public:
        int singlePath, maxPath;
        ResultType(int singlePath, int maxPath) {
            this->singlePath = singlePath;
            this->maxPath = maxPath;
        }
    };
    ResultType helper(TreeNode *root) {
        if (root == NULL) {
            return ResultType(0, INT_MIN);
        }
        //divide
        ResultType left = helper(root->left);
        ResultType right = helper(root->right);
        //conquer
        int singlePath = max(left.singlePath, right.singlePath) + root->val;
        singlePath = max(singlePath, 0);

        int maxPath = max(left.maxPath, right.maxPath);
        maxPath = max(maxPath, left.singlePath + right.singlePath + root->val);
        return ResultType(singlePath, maxPath);
    }
    int maxPathSum(TreeNode *root) {
        // write your code here
        ResultType result = helper(root);
        return result.maxPath;
    }
};
